gui/gui_state_container.h

```diff
@@ -0,0 +1,42 @@
+#ifndef GUI_STATE_CONTAINER_H
+#define GUI_STATE_CONTAINER_H
+
+#include "controller_history.h"
+#include <chrono>
+#include <mutex>
+
+// A thread-safe container for readable and writable state
+// of the GUI.
+class GuiStateContainer {
+public:
+  GuiStateContainer(std::chrono::milliseconds history_window)
+      : startup_time_(std::chrono::steady_clock::now()),
+        history_(history_window) {}
+
+  ControllerHistory::Instant GetStartupTime() { return startup_time_; }
+
+  void AppendHistory(const ControllerStatus &status) {
+    std::unique_lock<std::mutex> l(mu_);
+    history_.Append(std::chrono::steady_clock::now(), status);
+  }
+
+  GuiStatus GetGuiStatus() {
+    std::unique_lock<std::mutex> l(mu_);
+    return gui_status_;
+  }
+
+  std::vector<std::tuple<ControllerHistory::Instant, ControllerStatus>>
+  GetControllerStatusHistory() {
+    std::unique_lock<std::mutex> l(mu_);
+    return history_.GetHistory();
+  }
+
+private:
+  const ControllerHistory::Instant startup_time_;
+
+  std::mutex mu_;
+  GuiStatus gui_status_{};
```

Comment:
It doesn't have a constructor though. Please correct me if I'm wrong in the following reasoning:

The examples at https://en.cppreference.com/w/cpp/language/value_initialization imply that this will be zero-initialized (see `T1 t1{}`).

On the other hand, if I omit the `{}` then it's https://en.cppreference.com/w/cpp/language/default_initialization which calls the implicit default constructor, which per https://en.cppreference.com/w/cpp/language/default_constructor in this case will be a "trivial default constructor" that does nothing, i.e. does not zero-initialize.

Reply:
You're right, I forgot that GuiStatus is a proto.

The nanopb way of doing this is to use `GuiStatus_init_zero`.  Looking at the generated code, I don't think this is exactly same as `{}`, because enums are initialized to their lowest value, not 0.

Reply:
Used init_zero.